File: lib/iriq/synthesized_recognizer.rb
module Iriq
  # Recognizer built dynamically from a learned-prefix pattern.
  #
  # Used by Corpus#activate_proposal to promote a RecognizerProposal
  # into a live Recognizer that the classifier ensemble consults. Same
  # shape as the built-in Recognizers — uuid, date, integer — but the
  # pattern + type are supplied at construction instead of compiled-in.
  #
  #   r = SynthesizedRecognizer.new(prefix: "ghp_", type: :ghp)
  #   r.try("ghp_abcdef123")  # → {type: :ghp, confidence: 1.0, specificity: 1.0}
  #
  # Pattern: `<prefix><[A-Za-z0-9]+>` — anchored, alphanumeric suffix
  # only. Matches the same shape PrefixUnderscoreId proposes from, so
  # round-trip (propose → activate → reinfer) reclassifies the same
  # values the proposal was derived from.
  #
  # Specificity defaults to SEMANTIC. A learned prefix is very specific
  # by construction (a distinctive literal prefix that recurred enough
  # to clear the proposal noise floor) — calling it as confident as a
  # built-in UUID is reasonable.
  class SynthesizedRecognizer < Recognizer
    attr_reader :prefix, :type, :specificity

    def self.from_proposal(proposal)
      new(prefix: proposal.prefix, type: proposal.suggested_type)
    end

    def initialize(prefix:, type:, specificity: Specificity::SEMANTIC)
      raise ArgumentError, "prefix must be a non-empty string" if prefix.nil? || prefix.empty?
      raise ArgumentError, "type must be a symbol" unless type.is_a?(Symbol)

      @prefix      = prefix
      @type        = type
      @specificity = specificity
      @pattern     = /\A#{Regexp.escape(prefix)}[A-Za-z0-9]+\z/.freeze
    end

    def try(segment)
      return nil unless segment.start_with?(@prefix) && @pattern.match?(segment)

      { type: @type, confidence: 1.0, specificity: @specificity }
    end

    def to_dump
      { "prefix" => @prefix, "type" => @type.to_s, "specificity" => @specificity }
    end

    def self.from_dump(h)
      new(
        prefix:      h["prefix"],
        type:        h["type"].to_sym,
        specificity: h.fetch("specificity", Specificity::SEMANTIC),
      )
    end
  end
end

File: go/synthesized_recognizer.go
package iriq

import (
	"fmt"
	"regexp"
	"strings"
)

// SynthesizedRecognizer is a Recognizer built dynamically from a
// learned-prefix pattern.
//
// Used by Corpus.ActivateProposal to promote a RecognizerProposal into a
// live Recognizer that the classifier ensemble consults. Same shape as
// the built-in Recognizers (UUIDRecognizer, DateRecognizer,
// IntegerRecognizer) but the pattern + type are supplied at construction
// rather than compiled-in.
//
//   r := NewSynthesizedRecognizer("ghp_", "ghp", SpecificitySemantic)
//   r.Try("ghp_abcdef123")  // → {Type: ghp, Confidence: 1.0, Specificity: 1.0}, true
//
// Pattern: `<prefix><[A-Za-z0-9]+>` — anchored, alphanumeric suffix only.
// Matches the same shape PrefixUnderscoreIdStrategy proposes from, so
// round-trip (propose → activate → reinfer) reclassifies the same
// values the proposal was derived from.
type SynthesizedRecognizer struct {
	Prefix      string
	Type        SegmentType
	Specificity float64
	pattern     *regexp.Regexp
}

// NewSynthesizedRecognizer builds a Recognizer for the given prefix.
// Specificity defaults to SpecificitySemantic when zero — learned prefixes
// are very specific by construction.
func NewSynthesizedRecognizer(prefix, typeName string, specificity float64) *SynthesizedRecognizer {
	if prefix == "" {
		panic("SynthesizedRecognizer: prefix must be non-empty")
	}
	if specificity == 0 {
		specificity = SpecificitySemantic
	}
	pattern := regexp.MustCompile("^" + regexp.QuoteMeta(prefix) + "[A-Za-z0-9]+$")
	return &SynthesizedRecognizer{
		Prefix:      prefix,
		Type:        SegmentType(typeName),
		Specificity: specificity,
		pattern:     pattern,
	}
}

// SynthesizedRecognizerFromProposal mirrors the Ruby convenience
// constructor — pulls prefix + suggested_type out of the proposal.
func SynthesizedRecognizerFromProposal(p RecognizerProposal) *SynthesizedRecognizer {
	return NewSynthesizedRecognizer(p.Prefix, p.SuggestedType, SpecificitySemantic)
}

// Try implements Recognizer.
func (s *SynthesizedRecognizer) Try(segment string) (Verdict, bool) {
	if !strings.HasPrefix(segment, s.Prefix) || !s.pattern.MatchString(segment) {
		return Verdict{}, false
	}
	return Verdict{Type: s.Type, Confidence: 1.0, Specificity: s.Specificity}, true
}

// Dump returns the serialized form for persistence in a Corpus.
func (s *SynthesizedRecognizer) Dump() map[string]any {
	return map[string]any{
		"prefix":      s.Prefix,
		"type":        string(s.Type),
		"specificity": s.Specificity,
	}
}

// SynthesizedRecognizerFromDump reconstructs a Recognizer from the
// storage form produced by Dump().
func SynthesizedRecognizerFromDump(dump map[string]any) (*SynthesizedRecognizer, error) {
	prefix, ok := dump["prefix"].(string)
	if !ok || prefix == "" {
		return nil, fmt.Errorf("SynthesizedRecognizer: missing/invalid prefix in dump")
	}
	typeStr, _ := dump["type"].(string)
	if typeStr == "" {
		return nil, fmt.Errorf("SynthesizedRecognizer: missing type in dump")
	}
	spec, _ := dump["specificity"].(float64)
	return NewSynthesizedRecognizer(prefix, typeStr, spec), nil
}
